src/librustc_codegen_llvm/asm.rs

```diff
@@ -427,6 +428,11 @@ fn reg_to_llvm(reg: InlineAsmRegOrRegClass) -> String {
             | InlineAsmRegClass::Arm(ArmInlineAsmRegClass::qreg_low4) => "x",
             InlineAsmRegClass::Arm(ArmInlineAsmRegClass::dreg)
             | InlineAsmRegClass::Arm(ArmInlineAsmRegClass::qreg) => "w",
+            InlineAsmRegClass::Hexagon(HexagonInlineAsmRegClass::reg) => "r",
+            InlineAsmRegClass::Hexagon(HexagonInlineAsmRegClass::creg)
+            | InlineAsmRegClass::Hexagon(HexagonInlineAsmRegClass::preg) => unimplemented!("control, pred regs TODO"),
```

Comment:
Now what about the interesting case where control regs like `p3_0` / `c4` overlap predicate registers?  Do we need to define some control regs for the sake of the overlap logic but bar them from use as in/out?

Reply:
Only define predicate registers, not control registers.

Reply:
So the inline assembly doesn't need to know that there's aliasing?

Say there's inline assembly like so:

```
asm!("
c4 = r0
{} = r1", out(preg));


```All of the predicate registers are subregs of "c4".  AFAICT overlapping_regs only considers registers which are legal but it seems like we should give a friendly error here but can't, right?

Reply:
Another somewhat related case occurs when we try to packetize the instructions, which would get rejected by the assembler:

```
let p;
asm!("{{
{} = r1
c4 = r0
}}", out(preg) p);

```
For this we'd at least emit an error but it would be a fatal `LLVM ERROR` one right?

Reply:
Could I define `c4` as an `#error` and still use it in the overlapping_regs logic?

Reply:
In rustc we only care about inputs/outputs, we never actually parse the asm string itself. We only do string substitution to replace `{}` with register names and then leave any errors to the assembler.

> For this we'd at least emit an error but it would be a fatal LLVM ERROR one right?

We have a system for properly reporting assembler errors. This will not be displayed as an internal LLVM error.

> Could I define c4 as an #error and still use it in the overlapping_regs logic?

`overlapping_regs` is only used to check if multiple inputs/outputs are assigned to the same registers. We don't check the asm code itself.

Reply:
Gotcha, thanks for clarifying!